rfcs/wptserve_py3.md

```diff
@@ -0,0 +1,121 @@
+
+*   Getting bytes out: “encode” the return value with latin-1
+*   Passing bytes in: “decode” the byte sequence with latin-1
+
+
+## Options
+
+### [Preferred] Keep the semantics: always use byte sequences
+
+The preferred approach of the community is to keep a consistent and semantically correct encoding model. That is to **always use byte sequences: `str` in Python 2 and `bytes` in Python 3** (with some rare exceptions below).
+
+#### Detailed changes
+
+wptserve changes:
+
+*   Introduce a pair of helper functions `isomorphic_{encode,decode}` that use ISO-8859-1 under the hood and only encode/decode if necessary; both of them can accept either binary or text strings, but always return binary/text strings respectively regardless of the Python version.
+*   Most public APIs for custom handlers (including methods and properties of `Request` and `Response`) are binary, with the notable exception of the response body which can be text or binary strings.
+
+Guidelines for converting custom handlers:
+
+*   Make sure all strings are either always text or always bytes; all string literals in handlers should be prefixed with `b""` or `u""`. This will be enforced through a [lint](https://github.com/web-platform-tests/wpt/pull/23347). Existing code needs to be converted and reviewed manually ([example](https://github.com/web-platform-tests/wpt/pull/23363)). Note that we could add a hack to setters to accept both `str` and `bytes`, but on the getter side there is little we could do -- we don’t know whether the returned value will be used together with `bytes` or `str`, and things like `dict.get()` will be particularly tricky.
+*   **Headers** on both requests and responses are always *binary strings* for both keys and values. This includes:
+    * **HTTP Basic Auth**: `Request.auth.{username,password}` are binary strings.
+    * **Cookies**: Names and values in cookies are always binary strings, including `Request.cookies` and `Response.{set,unset,delete}_cookie`.
+*   **Request URL/form parameters** are all *binary strings* for both keys and values, accessible via `Request.GET` or `Request.POST`.
+*   **Response body** can be written via `Response.writer.*` or via the [return values](https://web-platform-tests.org/tools/wptserve/docs/handlers.html#python-handlers), which can be either text or binary strings, but the two types should never be mixed and string literals must be prefixed.
+    * As an exception, `Response.set_error` takes only text message strings.
```

Comment:
Can we briefly summarize why? (If not brief, then don't)

Reply:
```suggestion
    * As an exception, `Response.set_error` takes only text message strings, as this message is used in the JSON response.
```